crates/turborepo-scm/src/package_deps.rs

```diff
@@ -207,6 +242,45 @@ impl Git {
         hash_objects(&self.root, &full_pkg_path, to_hash, &mut hashes)?;
         Ok(hashes)
     }
+
+    #[tracing::instrument(skip(self, turbo_root, inputs))]
+    fn get_package_file_hashes_from_inputs_and_index<S: AsRef<str>>(
+        &self,
+        turbo_root: &AbsoluteSystemPath,
+        package_path: &AnchoredSystemPath,
+        inputs: &[S],
+        include_configs: bool,
+    ) -> Result<GitHashes, Error> {
+        // collect the default files and the inputs
+        let default_file_hashes =
+            self.get_package_file_hashes_from_index(turbo_root, package_path)?;
+
+        // we need to get hashes for excludes separately so we can remove them from the
+        // defaults later on
+        let mut includes = Vec::new();
+        let mut excludes = Vec::new();
+        for input in inputs {
+            let input_str = input.as_ref();
+            if let Some(exclude) = input_str.strip_prefix('!') {
+                excludes.push(exclude);
+            } else {
+                includes.push(input_str);
+            }
+        }
+        let manual_includes_hashes =
+            self.get_package_file_hashes_from_inputs(turbo_root, package_path, &includes, true)?;
+        let manual_excludes_hashes =
+            self.get_package_file_hashes_from_inputs(turbo_root, package_path, &excludes, false)?;
```

Comment:
I think we should add a check if `includes`/`excludes` are empty before performing these calls as they'll end up doing some amount of throw away work.

Reply:
Good call, we can do this with `excludes`, but with the way things are currently structured - if you manually supply `inputs`, we will _always_ add `package.json` and `turbo.json` to them, meaning the includes are _never_ empty.